### ISSUE TITLE
[Feature] SGLang llguidance Structured Output Backend NPU Support

### ISSUE THREAD

**Issue by loading66** [Ascend/sglang#1107]
## Task Name
SGLang llguidance Structured Output Backend NPU Support

### Task Description
SGLang's structured output and constrained function calling modes rely on grammar backends to apply token bitmasks to logits before sampling. Among existing backends, the llguidance backend is implemented for CUDA and cannot run correctly on Ascend NPU. This task aims to add an NPU device branch to the llguidance backend's bitmask application path, enabling --grammar-backend llguidance to work correctly on Ascend NPU for structured output and constrained function calling, thereby closing the structured output backend gap between NPU and GPU.

The goal is to achieve inference adaptation on the Ascend NPU platform and to verify architectural compatibility and inference accuracy.
Development platform: Atlas 800T A2 or A3

### Acceptance Criteria

With --grammar-backend llguidance on Ascend NPU, JSON Schema, regex, EBNF, and structural_tag constraints all generate correctly, and outputs strictly satisfy the constraints.

Behavior with the llguidance backend on NPU is consistent with GPU.

Add corresponding unit tests.

**Deliverables**

This phase extends the existing framework/repository and requires development code and documentation. The main development points are:

1. Submit a PR
  - Complete inference adaptation code.
  - Add test cases under `test/registered/ascend` to guard accuracy.
  - Provide a full PR description that includes accuracy and performance results, following community conventions.
  - (Optional) Supplement `docs_new/docs/hardware-platforms/ascend-npus/ascend_npu_faq.mdx` with problems encountered and solved during adaptation.
2. Update the Issue
  - The issue may include development experience, problems encountered, and their solutions.

**Acceptance Criteria**

1. Accuracy / Performance requirements
- Accuracy: The adapted model's accuracy on representative evaluation tasks must not deviate more than 1% from the official baseline.
- Performance: N/A

2. The PR must contain complete inference code, test code, and a PR description.

### PR Merge
After local testing and verification, create a PR against the `main` branch of https://github.com/sgl-project/sglang.

### Point of Contact
@amote-i

Welcome to the community and thank you for your contribution 🎉!

**Comment by LongCatchen** [comment]
Claim this task.

**Plan:** Add NPU device branch to llguidance grammar backend bitmask path; unit tests under `test/registered/ascend`.

**Branch:** `LongCatchen/sglang` → `feature/issue-1107-llguidance-npu`

**Env:** 1×910B3 smoke optional (Qwen2.5-7B)

**Harness:** H2 dispatched for implementation.

**Comment by LongCatchen** [comment]
PR submitted: https://github.com/sgl-project/sglang/pull/36074

Branch: `LongCatchen/sglang` → `feature/issue-1107-llguidance-npu`

**Comment by LongCatchen** [comment]
## NPU local retest — PASS (2026-08-23)

**Environment:** Atlas 800T A2, 8×910B3, Docker `sglang_dev` (`quay.io/ascend/sglang:cann9.0.0-910b-v0.5.16`), NPU test host

**Commit:** `f45f490` — fix: keep llguidance NPU mock/reference masking on CPU tensors

**Command:**
```bash
export PYTHONPATH=python
python3 -m pytest test/registered/ascend/test_llguidance_npu_mask.py -q --tb=short
```

**Result:** 4 passed in ~14s

**Upstream PR:** https://github.com/sgl-project/sglang/pull/36074

**Fix notes:** NPU mock applies bitmask on CPU tensors to avoid device mismatch when logits report `npu:0` via subclass.

**Comment by LongCatchen** [comment]
## NPU end-to-end retest — PASS (2026-09-21)

Same PR branch `f45f490`, run on an NPU test host with that checkout (not the image tree).

**Unit test:** `test/registered/ascend/test_llguidance_npu_mask.py` — 4 passed, 15 warnings in 37.96s.

**Server:** `Qwen/Qwen2.5-0.5B-Instruct` with `--device npu --attention-backend ascend --grammar-backend llguidance`. Startup log shows `grammar_backend=llguidance` and `The server is fired up and ready to roll!`.

**Constrained request:** `POST /v1/chat/completions` with JSON Schema requiring `name` (string) and `population` (integer). Response:

```json
{
  "name": "Paris",
  "population": 2184000
}
```

`finish_reason=stop`, `completion_tokens=23`.

Upstream note: https://github.com/sgl-project/sglang/pull/36074#issuecomment-5756228085

**Comment by LongCatchen** [comment]
Follow-up NPU retest for Ascend/sglang #1107 acceptance (JSON Schema / regex / EBNF / structural_tag), on a **single NPU** (`/dev/davinci0` only, `ASCEND_RT_VISIBLE_DEVICES=0`), branch `f45f490`.

### HTTP constrained generate

| Constraint | Result |
|---|---|
| `response_format.json_schema` | `{"name":"Paris","population":2184000}` |
| `regex: (Paris\|London)` | `Paris` |
| `ebnf: root ::= "yes" \| "no"` | `yes` |
| `structural_tag` (legacy structures/triggers) | Grammar compiles with real tokenizer special tokens (`<\|box_start\|>` / `<\|box_end\|>`). Free text before the trigger is allowed by the structural_tag grammar (`at_least_one=false`). |

### structural_tag NPU bitmask (after trigger)

After `accept_token(<|box_start|>)`, `GuidanceGrammar.apply_vocab_mask` on NPU logits keeps **7** allowed tokens, all `{` / `{\n` / `{"` variants — i.e. the JSON object start required by the schema. This exercises `torch.ops.npu.apply_token_bitmask`.

### Unit test

```
4 passed, 15 warnings in 37.96s
```

(`test/registered/ascend/test_llguidance_npu_mask.py`)
